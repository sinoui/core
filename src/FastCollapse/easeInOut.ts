/* eslint-disable no-return-assign */
/* eslint-disable no-param-reassign */
/* eslint-disable no-cond-assign */

/**
 * 缓入缓出
 *
 * @param elapsed 已走完的时间
 * @param initialValue 初始值
 * @param amountOfChange 总的值变化量
 * @param duration 时长
 */
export default function easeInOut(
  elapsed: number,
  initialValue: number,
  amountOfChange: number,
  duration: number,
): number {
  if ((elapsed /= duration / 2) < 1) {
    return (amountOfChange / 2) * elapsed * elapsed * elapsed + initialValue;
  }
  return (
    (amountOfChange / 2) * ((elapsed -= 2) * elapsed * elapsed + 2) +
    initialValue
  );
}
