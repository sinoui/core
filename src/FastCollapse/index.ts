import FastCollapse from './FastCollapse';

export default FastCollapse;
