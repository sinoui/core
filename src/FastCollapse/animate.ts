import easeInOut from './easeInOut';

/**
 * 过渡动画
 *
 * @param duration 动画时长
 * @param frameCallback 每一帧都会执行的回调函数
 */
export default function animate(
  duration: number,
  frameCallback: (getValue: (start: number, end: number) => number) => void,
) {
  const startTime = Date.now();
  let rafId = -1;

  const tick = () => {
    const elapsed = Math.min(duration, Date.now() - startTime);
    frameCallback((start: number, end: number) =>
      easeInOut(elapsed, start, end - start, duration),
    );

    if (elapsed < duration) {
      rafId = requestAnimationFrame(tick);
    } else {
      rafId = -1;
    }
  };

  tick();

  return () => {
    if (rafId !== -1) {
      cancelAnimationFrame(rafId);
      rafId = -1;
    }
  };
}
