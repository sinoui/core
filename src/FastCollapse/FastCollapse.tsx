/* eslint-disable no-param-reassign */
import React, { useRef, useEffect, useLayoutEffect } from 'react';
import { Transition } from 'react-transition-group';
import styled from 'styled-components';
import transitions from '@sinoui/theme/transitions';
import type { TransitionStatus } from 'react-transition-group/Transition';
import animate from './animate';
import type BaseTransitionProps from '../transitions/BaseTransitionProps';
import getDuration from '../transitions/getDuration';
import type { TransitionTimeout } from '../transitions/getDuration';
import useMultiRefs from '../utils/useMultiRefs';

interface CollapseWrapperProps {
  $hidden?: boolean;
}

const CollapseWrapper = styled.div<CollapseWrapperProps>`
  overflow: hidden;
  visibility: ${({ $hidden }) => ($hidden ? 'hidden' : '')};
  transform-origin: top left;
  backface-visibility: hidden; /* 兼容 IE11，启用GPU加速 */
  will-change: transform;

  & > div {
    transform-origin: top left;
    backface-visibility: hidden; /* 兼容 IE11，启用GPU加速 */
    will-change: transform;
  }
`;

type Props = Omit<BaseTransitionProps, 'timeout'> & {
  /**
   * 压缩后的高度。默认为`0`。单位为`px`。
   */
  collapsedHeight?: number;
  /**
   * 压缩后的宽度。默认为`0`。单位为`px`。
   */
  collapsedWidth?: number;
  /**
   * 指定动画时长。如果指定为`auto`，则会根据实际高度来计算时长。
   */
  timeout?: TransitionTimeout;
  /**
   * 收缩方向，`both`表示垂直和水平方向同时收缩
   */
  direction?: 'vertical' | 'horizontal' | 'both';
};

/**
 * 压缩元素的过渡效果。
 *
 * FastCollapse组件适合用于弹窗的动效。如果需要压缩的是布局中的元素，则需要使用Collapse组件。
 *
 * 实现思路参考： https://developers.google.com/web/updates/2017/03/performant-expand-and-collapse
 */
const Collapse = React.forwardRef<HTMLDivElement, Props>(function Collapse(
  props,
  ref,
) {
  const {
    children,
    timeout = transitions.duration.standard,
    in: inProp,
    collapsedHeight = 0,
    collapsedWidth = 0,
    onEnter,
    onExit,
    direction = 'vertical',
    ...rest
  } = props;
  const contentRef = useRef<HTMLElement>(null);
  const handleContentRef = useMultiRefs(contentRef, ref);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const autoTimeout = useRef<number>(0);
  const timerRef = useRef<number>();
  const cancelAnimateRef = useRef<Function | null>(null);
  const inRef = useRef<boolean>(!!inProp);

  /**
   * 处理动画
   *
   * @param status 状态
   */
  const handleAnimate = (status: 'enter' | 'exit') => {
    const node = contentRef.current;
    const content = wrapperRef.current;
    if (!node || !content) {
      return;
    }
    const duration = getDuration(timeout, status, content);
    autoTimeout.current = timeout === 'auto' ? duration : 0;
    const initX =
      direction === 'vertical' ? 1 : collapsedWidth / content.clientWidth;
    const initY =
      direction === 'horizontal' ? 1 : collapsedHeight / content.clientHeight;
    const start = {
      x: status === 'enter' ? initX : 1,
      y: status === 'enter' ? initY : 1,
    };
    const end = {
      x: status === 'enter' ? 1 : initX,
      y: status === 'enter' ? 1 : initY,
    };

    // eslint-disable-next-line no-unused-expressions
    cancelAnimateRef.current?.();

    cancelAnimateRef.current = animate(duration, (interpolation) => {
      const scaleX =
        direction === 'vertical' ? 1 : interpolation(start.x, end.x);
      const scaleY =
        direction === 'horizontal' ? 1 : interpolation(start.y, end.y);
      const invScaleX = scaleX === 0 ? 60 : 1 / scaleX;
      const invScaleY = scaleY === 0 ? 60 : 1 / scaleY;
      node.style.transform = `scale(${scaleX}, ${scaleY})`;
      content.style.transform = `scale(${invScaleX}, ${invScaleY})`;
    });
  };

  const handleEnter: any = (isAppearing: boolean) => {
    handleAnimate('enter');

    const node = contentRef.current;
    if (onEnter && node) {
      onEnter(node, isAppearing);
    }
  };

  const handleExit = () => {
    handleAnimate('exit');

    const node = contentRef.current;
    if (onExit && node) {
      onExit(node);
    }
  };

  const addEndListener: any = (done: () => void) => {
    timerRef.current = setTimeout(done, autoTimeout.current);
  };

  useEffect(() => {
    return () => window.clearTimeout(timerRef.current);
  }, []);

  useLayoutEffect(() => {
    if (!inRef.current) {
      const node = contentRef.current;
      const content = wrapperRef.current;
      if (!node || !content) {
        return;
      }

      const scaleX =
        direction === 'vertical' ? 1 : collapsedWidth / content.clientWidth;
      const scaleY =
        direction === 'horizontal' ? 1 : collapsedHeight / content.clientHeight;

      const invScaleX = scaleX === 0 ? 60 : 1 / scaleX;
      const invScaleY = scaleY === 0 ? 60 : 1 / scaleY;

      node.style.transform = `scale(${scaleX}, ${scaleY})`;
      content.style.transform = `scale(${invScaleX}, ${invScaleY})`;
    }
  }, [collapsedHeight, collapsedWidth, direction]);

  return (
    <Transition
      appear
      in={inProp}
      timeout={timeout === 'auto' ? undefined : timeout}
      onEnter={handleEnter}
      onExit={handleExit}
      addEndListener={(timeout === 'auto' ? addEndListener : undefined) as any}
      nodeRef={wrapperRef}
      {...rest}
    >
      {(status: TransitionStatus, childProps: any) => (
        <CollapseWrapper
          className="sinoui-collapse-container"
          $hidden={status === 'exited' && !inProp && collapsedHeight === 0}
          ref={handleContentRef}
          {...childProps}
        >
          <div ref={wrapperRef} className="sinoui-collapse-content">
            {children}
          </div>
        </CollapseWrapper>
      )}
    </Transition>
  );
});

export default Collapse;
