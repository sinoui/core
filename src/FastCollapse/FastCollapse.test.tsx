import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import 'jest-styled-components';
import { create } from 'react-test-renderer';
import animate from './animate';
import FastCollapse from './FastCollapse';

jest.mock('./animate');

afterEach(cleanup);

/**
 * 模拟动画内容元素的尺寸
 *
 * @param collapseElement 动画根元素
 * @param width 宽度
 * @param height 高度
 */
const mockCollapseContentSize = (
  collapseElement: HTMLDivElement | null,
  width = 100,
  height = 100,
) => {
  const contentElement = collapseElement?.querySelector(
    '.sinoui-collapse-content',
  );
  if (contentElement) {
    Object.defineProperty(contentElement, 'clientWidth', {
      configurable: true,
      value: width,
    });
    Object.defineProperty(contentElement, 'clientHeight', {
      configurable: true,
      value: height,
    });
  }
};

beforeEach(() => {
  (animate as jest.Mock).mockImplementation(
    (_duration: number, frameCallback) => {
      const interpolation = (_startValue: number, endValue: number) => endValue;
      frameCallback(interpolation);
    },
  );
});

it('默认展现', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse data-testid="collapse" in>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transform: scale(1, 1)');
  expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
    'transform: scale(1, 1)',
  );
});

it('默认隐藏', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse
        data-testid="collapse"
        in={false}
        ref={mockCollapseContentSize}
      >
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transform: scale(1, 0)');
  expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
    'transform: scale(1, 60)',
  );
});

it('隐藏 -> 展现', () => {
  const { getByTestId, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse data-testid="collapse" in={false}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  rerender(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse data-testid="collapse" in>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transform: scale(1, 1)');
  expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
    'transform: scale(1, 1)',
  );
});

it('显示 -> 隐藏', () => {
  const { getByTestId, rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse data-testid="collapse" in ref={mockCollapseContentSize}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  rerender(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse
        data-testid="collapse"
        in={false}
        ref={mockCollapseContentSize}
      >
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(getByTestId('collapse')).toHaveStyle('transform: scale(1, 0)');
  expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
    'transform: scale(1, 60)',
  );
});

it('clean timeout after unmounted', () => {
  const onEntered = jest.fn();
  const { unmount } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse
        in
        timeout="auto"
        data-testid="collapse"
        onEntered={onEntered}
      >
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  unmount();
  expect(onEntered).not.toBeCalled();
});

it('ref', () => {
  const ref = React.createRef<HTMLDivElement>();
  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse in timeout="auto" data-testid="collapse" ref={ref}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(ref.current).toBe(getByTestId('collapse'));
});

it('onEnter, onEntering', () => {
  const onEnter = jest.fn();
  const onEntering = jest.fn();

  render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse in onEnter={onEnter} onEntering={onEntering}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(onEnter).toBeCalled();
  expect(onEntering).toBeCalled();
});

it('onExit, onExiting', () => {
  const onExit = jest.fn();
  const onExiting = jest.fn();
  const { rerender } = render(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse in onExit={onExit} onExiting={onExiting}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );
  rerender(
    <ThemeProvider theme={defaultTheme}>
      <FastCollapse in={false} onExit={onExit} onExiting={onExiting}>
        <div>测试</div>
      </FastCollapse>
    </ThemeProvider>,
  );

  expect(onExit).toBeCalled();
  expect(onExiting).toBeCalled();
});

describe('水平方向', () => {
  it('默认隐藏', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in={false}
          ref={mockCollapseContentSize}
          direction="horizontal"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    expect(getByTestId('collapse')).toHaveStyle('transform: scale(0, 1)');
    expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
      'transform: scale(60, 1)',
    );
  });

  it('显示 -> 隐藏', () => {
    const { getByTestId, rerender } = render(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in
          ref={mockCollapseContentSize}
          direction="horizontal"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    rerender(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in={false}
          ref={mockCollapseContentSize}
          direction="horizontal"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    expect(getByTestId('collapse')).toHaveStyle('transform: scale(0, 1)');
    expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
      'transform: scale(60, 1)',
    );
  });
});

describe('双方向', () => {
  it('默认隐藏', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in={false}
          ref={mockCollapseContentSize}
          direction="both"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    expect(getByTestId('collapse')).toHaveStyle('transform: scale(0, 0)');
    expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
      'transform: scale(60, 60)',
    );
  });

  it('显示 -> 隐藏', () => {
    const { getByTestId, rerender } = render(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in
          ref={mockCollapseContentSize}
          direction="both"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    rerender(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse
          data-testid="collapse"
          in={false}
          ref={mockCollapseContentSize}
          direction="both"
        >
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    );

    expect(getByTestId('collapse')).toHaveStyle('transform: scale(0, 0)');
    expect(getByTestId('collapse').querySelector('div')).toHaveStyle(
      'transform: scale(60, 60)',
    );
  });
});

describe('快照测试', () => {
  it('展现', () => {
    const tree = create(
      <ThemeProvider theme={defaultTheme}>
        <FastCollapse data-testid="collapse" in>
          <div>测试</div>
        </FastCollapse>
      </ThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
