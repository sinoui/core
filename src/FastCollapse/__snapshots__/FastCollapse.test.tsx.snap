// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 展现 1`] = `
.c0 {
  overflow: hidden;
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  will-change: transform;
}

.c0 > div {
  -webkit-transform-origin: top left;
  -ms-transform-origin: top left;
  transform-origin: top left;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  will-change: transform;
}

<div
  className="c0 sinoui-collapse-container"
  data-testid="collapse"
>
  <div
    className="sinoui-collapse-content"
  >
    <div>
      测试
    </div>
  </div>
</div>
`;
